### PR TITLE
[FEAT] EVA: ensure deterministic behavior of SVD on multi gpu setups

### PR DESCRIPTION
**Pull request by sirluk**
Deterministic behavior of SVD computations on layer inputs is important on multi gpu setups to ensure resulting EVA state dicts are identical. Currently deterministic behaviour of SVD is ensured by just setting torch.manual_seed() inside `_get_eva_state_dict` which is not ideal because it can interfere with the seed set by the user. I therefore moved this functionality inside the `IncrementalPCA` class.

I also added a new argument for `get_eva_state_dict`: `gather_distributed_inputs` which is `False` by default and defines if inputs should be concatenated for the SVD computation for distributed runs.

### CODE REVIEW

File: src/peft/tuners/lora/eva.py [hunk 1]
```diff
@@ -43,8 +43,14 @@ class _Hook:
     A base class for hooks that prepares layer inputs for EVA.
     """
 
-    def __init__(self, name: str, prepare_layer_inputs_fn: Optional[callable] = None):
+    def __init__(
+        self,
+        name: str,
+        prepare_layer_inputs_fn: Optional[callable] = None,
+        gather_distributed_inputs: bool = False,
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/eva.py:50]: For my understanding, setting `gather_distributed_inputs=True` in a non-distributed setting makes no sense. In what circumstances would we `gather_distributed_inputs=False` in a distributed setting?

File: src/peft/tuners/lora/eva.py [hunk 2]
```diff
@@ -314,6 +317,14 @@ def _get_rank_distribution(hooks, layer_hook_map, equal_inputs_map, rank_budget,
     if len(dataloader) == 0:
         raise ValueError("dataloader is empty")
 
+    # check if dist is initialized
+    if dist.is_initialized() and gather_distributed_inputs:
+        warnings.warn(
+            "torch.distributed is initialized and `gather_distributed_inputs` is True, "
+            "therefore EVA initialization will gather tensors from all ranks. "
+            "Ensure the model does not recieve the same inputs on different ranks."
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/eva.py:325]:
```suggestion
            "Ensure the model does not receive the same inputs on different ranks."
```

File: src/peft/tuners/lora/eva.py [hunk 3]
```diff
@@ -579,6 +597,7 @@ def get_eva_state_dict(
             case model_inputs is the mask used to determine which indices should be used for SVD (created by
             `prepare_model_inputs_fn_language_modeling`).
         adapter_name (str): The name of the adapter to compute the SVD for.
+        gather_distributed_inputs (bool): Whether to gather the layer inputs from all ranks. Default is False.
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/eva.py:600]: Let's add: "Set this to True if ...".

File: src/peft/tuners/lora/eva.py [hunk 4]
```diff
@@ -672,6 +693,7 @@ def initialize_lora_eva_weights(
             case model_inputs is the mask used to determine which indices should be used for SVD (created by
             `prepare_model_inputs_fn_language_modeling`).
         adapter_name (str): The name of the adapter to initialize the weights for.
+        gather_distributed_inputs (bool): Whether to gather the layer inputs from all ranks. Default is False.
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/eva.py:696]: Let's add: "Set this to True if ...".

File: src/peft/utils/incremental_pca.py [hunk 5]
```diff
@@ -66,11 +68,18 @@ def __init__(
             if lowrank_q < n_components:
                 raise ValueError("lowrank_q must be greater than or equal to n_components.")
 
-            def svd_fn(X):
-                U, S, V = torch.svd_lowrank(X, q=lowrank_q, niter=lowrank_niter)
+            def svd_fn(X, seed):
+                if seed is not None:
```
Comment (BenjaminBossan) [inline src/peft/utils/incremental_pca.py:72]: I haven't used it, but could this be a good situation to use [`fork_rng`](https://pytorch.org/docs/stable/random.html#torch.random.fork_rng)?

File: src/peft/utils/incremental_pca.py [hunk 6]
```diff
@@ -66,11 +68,18 @@ def __init__(
             if lowrank_q < n_components:
                 raise ValueError("lowrank_q must be greater than or equal to n_components.")
 
-            def svd_fn(X):
-                U, S, V = torch.svd_lowrank(X, q=lowrank_q, niter=lowrank_niter)
+            def svd_fn(X, seed):
```
Comment (BenjaminBossan) [inline src/peft/utils/incremental_pca.py:71]: Would it make sense to turn `svd_fn` into a proper method, with all arguments stored as `self` attributes (so `self.lowrank = lowrank` etc.). Then we don't need to `partial` over the arguments. Furthermore, this is more in line with sklearn: Save all the class arguments as `self` attributes and don't put any further logic inside of `__init__` -- this is because this logic is not executed when running `set_params` on the object.

Of course, this `IncrementalPCA` class is not a proper sklearn `BaseEstimator` with `get_params` and `set_params`, so it doesn't really matter, but it could still be nice for possible future integration to keep with the sklearn standards.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for these small fixes for distributed EVA. Overall, this looks good as is but I have added some comments for potential improvement, please check.

File: src/peft/tuners/lora/eva.py (hunk 1 above)
Comment (sirluk) [reply src/peft/tuners/lora/eva.py:50]: My thinking when adding this parameter was that users might use a distributed environment but for some reason do not create a distributed dataloader for EVA. For those cases we would have the same examples as layer inputs and thus when gathering we would have duplicates in the input to the SVD computation. Not sure if this is a valid scenario of if we can disregard this.

File: src/peft/utils/incremental_pca.py (hunk 5 above)
Comment (sirluk) [reply src/peft/utils/incremental_pca.py:72]: I didnt know about this function but it looks useful. I will check if I can implement it.

File: src/peft/utils/incremental_pca.py (hunk 6 above)
Comment (sirluk) [reply src/peft/utils/incremental_pca.py]: done, I cleaned up the __init__ method and defined the svd methods as proper class methods

File: src/peft/tuners/lora/eva.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/eva.py:50]: Ah yes, that makes sense. Could you please add this info to the docstring below, that way users know when they should set this in a distributed setting.

One thing I wonder: Could we make this `Optional[bool]` and if the value is `None` and the setting is distributed, automatically switch to `True`, otherwise to `False`? This is probably the most common case. Users still can pass `False` explicitly in a distributed setting if they want to.

Comment (sirluk) [reply src/peft/tuners/lora/eva.py:50]: I think we can leave it as `bool` and set the default to `True` as the Hooks anyway only gather if `dist.is_initialized()`. So if the setting is not distributed this argument has no effect.

Comment (BenjaminBossan) [inline src/peft/tuners/lora/eva.py:50]: Since `gather_distributed_inputs` defaults to `True` now in `initialize_lora_eva_weights`, should it also be changed here?

File: examples/eva_finetuning/eva_finetuning_multi_gpu.py [hunk 7]
```diff
@@ -96,7 +96,8 @@
 )
 
 # EVA initialization
-eva_state_dict = get_eva_state_dict(model, dataloader, peft_config)
+# It is important to set `gather_distributed_inputs=True` here if you use a distributed data sampler.
+eva_state_dict = get_eva_state_dict(model, dataloader, peft_config, gather_distributed_inputs=True)
```
Comment (BenjaminBossan) [inline examples/eva_finetuning/eva_finetuning_multi_gpu.py:100]:
```suggestion
eva_state_dict = get_eva_state_dict(model, dataloader, peft_config)
```

As `True` is the default.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the updates. I have some nits, WDYT?

**Comment by sirluk** [comment]
@BenjaminBossan thanks for the suggestions, I integrated them!

**Review by BenjaminBossan — APPROVED** [review]
Thanks for these updates to EVA, LGTM.
